### PR TITLE
[Fix] Incorrect Labels Shift for Packed Data

### PR DESCRIPTION
**Pull request by pbontrager**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

@felipemello1 discovered that how we compute our shifted labels in the [recipe](https://github.com/pytorch/torchtune/blob/1075d9cddfd28d5f1d4ab4e154af6fa80d537fa5/recipes/full_finetune_single_device.py#L642) breaks when we do packed training. Instead of shifting the whole sequence over one and then appending an ignore token at the end of the whole packed sequence, we need to do it on the individual sequences before packing.

This would only affect a single token at each boundary, so it's unclear how much it affected performance in practice. 

#### Changelog
What are the changes made in this PR?
* Updated 10 sft based recipes `git grep hstack -> 10 recipes (full, lora, qat, kd, dev)`
* For each recipe removed ignore_index_cache and label shift logic
* Updated SFTDataset and TextCompletionDataset to handle shifting the labels
* Updated 7 dataset unit tests to have correct exepected_labels

#### Test plan
- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

To compare loss I ran `tune run full_finetune_single_device --config llama3_2/1B_full_single_device tokenizer.max_seq_len=8092 dataset.packed=True compile=True optimizer=torch.optim.AdamW metric_logger=torchtune.training.metric_logging.WandBLogger metric_logger.project="Fix Packed"` on main and on this branch.

Screenshot 2025-04-18 at 4 58 33 PM

The bug doesn't seem to have made a noticeable impact on Alpaca loss.

### CODE REVIEW

**Review by joecummings — APPROVED** [review]
unblockdd
